File: clients/sdk/src/pay.ts
// Atomic spend + record_payment in one transaction. The agent signs the
// spend half (credit_vault), the service keypair signs the reputation
// half (record_payment). Both land or neither does — no half-states
// where USDC moved but reputation didn't.
//
// Mirrors `Spender::pay()` in `clients/runtime/src/lib.rs`.

import { AnchorError, BN } from "@coral-xyz/anchor";
import {
  Connection,
  type Keypair,
  PublicKey,
  sendAndConfirmTransaction,
  SystemProgram,
  Transaction,
} from "@solana/web3.js";
import {
  TOKEN_PROGRAM_ID,
  createAssociatedTokenAccountIdempotentInstruction,
  getAssociatedTokenAddress,
} from "./constants.js";
import { decodeCreditVault, decodeSpendPolicy } from "./decoders.js";
import {
  AccountNotFoundError,
  HourlyLimitExceededError,
  LifetimeLimitExceededError,
  NotWhitelistedError,
  PerTxLimitExceededError,
  ReceiptAlreadyRecordedError,
  RecordPaymentError,
  ServiceInactiveError,
  VaultFrozenError,
  ZeroAmountError,
} from "./errors.js";
import { guardSpend } from "./guardrails.js";
import {
  agentProfilePda,
  agentServiceLinkPda,
  policyPda,
  receiptUsedPda,
  serviceRegistryPda,
  toPubkey,
  vaultPda,
} from "./pda.js";
import {
  buildProvider,
  creditVaultProgram,
  reputationProgram,
} from "./programs.js";
import type { CreditVaultAccount, Pubkeyish, SpendPolicyAccount } from "./types.js";

export type PayArgs = {
  /** Agent keypair — spends from the vault and pays tx fees. */
  agent: Keypair;
  /** Service keypair — co-signs the reputation half. Its pubkey must
   *  match the service registered on chain. */
  service: Keypair;
  /** Vault owner. */
  owner: Pubkeyish;
  /** Amount in micro-USDC. */
  amountUsdc: number;
  /** 32-byte receipt hash. Common: `sha256(agent|service|tick|...)`. */
  receiptHash: Uint8Array;
  connection: Connection;
};

export type PayResult = {
  signature: string;
};

/**
 * Atomic spend + record_payment. The vault burn and reputation accrual
 * happen in one transaction — without this, a service could mirror only
 * payments it actually received, but the agent's score wouldn't reflect
 * its own spending. Both sides of the SDK need this primitive to keep
 * reputation honest.
 */
export async function pay(args: PayArgs): Promise<PayResult> {
  const { agent, service, connection, amountUsdc } = args;
  const owner = toPubkey(args.owner);

  if (amountUsdc <= 0) {
    throw new ZeroAmountError();
  }
  if (args.receiptHash.length !== 32) {
    throw new Error(
      `pay: receiptHash must be 32 bytes, got ${args.receiptHash.length}`,
    );
  }

  const vaultAddr = vaultPda(owner, agent.publicKey);
  const policyAddr = policyPda(vaultAddr);

  // Pre-flight against the local guardrail — bails on obvious policy
  // violations before we burn a tx slot. Mirrors `spend()` in client.ts.
  const provider = buildProvider(connection, agent);
  const cv = creditVaultProgram(provider);
  const rep = reputationProgram(provider);

  const [rawVault, rawPolicy, currentSlot] = await Promise.all([
    cv.account.creditVault.fetchNullable(vaultAddr),
    cv.account.spendPolicy.fetchNullable(policyAddr),
    connection.getSlot(),
  ]);
  if (!rawVault) throw new AccountNotFoundError(vaultAddr.toBase58());
  if (!rawPolicy) throw new AccountNotFoundError(policyAddr.toBase58());
  const vault = decodeCreditVault(vaultAddr, rawVault);
  const policy = decodeSpendPolicy(policyAddr, rawPolicy);
  guardSpend({
    vault,
    policy,
    service: service.publicKey,
    amountUsdc,
    currentSlot,
  });

  const serviceTokenAccount = getAssociatedTokenAddress(
    vault.usdc_mint,
    service.publicKey,
  );
  const createAtaIx = createAssociatedTokenAccountIdempotentInstruction(
    agent.publicKey,
    serviceTokenAccount,
    service.publicKey,
    vault.usdc_mint,
  );

  const agentProfile = agentProfilePda(agent.publicKey);
  const serviceRegistry = serviceRegistryPda(service.publicKey);
  const link = agentServiceLinkPda(agentProfile, serviceRegistry);
  const receipt = receiptUsedPda(args.receiptHash);

  // `.instruction()` is a stable Anchor API but isn't on our typed
  // builder wrappers (which only expose `.rpc()` for the typical happy
  // path). Cast at the boundary to compose multi-program txs ourselves.
  type InstructionBuilder = {
    instruction(): Promise<import("@solana/web3.js").TransactionInstruction>;
  };
  const spendBuilder = cv.methods.spend(new BN(amountUsdc)).accounts({
    agent: agent.publicKey,
    vault: vaultAddr,
    policy: policyAddr,
    vaultTokenAccount: vault.vault_token_account,
    serviceTokenAccount,
    tokenProgram: TOKEN_PROGRAM_ID,
  }) as unknown as InstructionBuilder;
  const recordBuilder = rep.methods
    .recordPayment(new BN(amountUsdc), args.receiptHash)
    .accounts({
      service: service.publicKey,
      agentProfile,
      serviceRegistry,
      agentServiceLink: link,
      receiptUsed: receipt,
      systemProgram: SystemProgram.programId,
    }) as unknown as InstructionBuilder;
  const spendIx = await spendBuilder.instruction();
  const recordIx = await recordBuilder.instruction();

  const tx = new Transaction().add(createAtaIx, spendIx, recordIx);
  try {
    const signature = await sendAndConfirmTransaction(
      connection,
      tx,
      [agent, service],
      { commitment: "confirmed" },
    );
    return { signature };
  } catch (err) {
    throw mapPayError(err, {
      service: service.publicKey,
      amountUsdc,
      vault,
      policy,
      receiptHash: args.receiptHash,
    });
  }
}

function mapPayError(
  err: unknown,
  ctx: {
    service: PublicKey;
    amountUsdc: number;
    vault: CreditVaultAccount;
    policy: SpendPolicyAccount;
    receiptHash: Uint8Array;
  },
): unknown {
  const message = err instanceof Error ? err.message : String(err);
  if (/already in use/i.test(message)) {
    return new ReceiptAlreadyRecordedError(ctx.receiptHash);
  }
  if (!(err instanceof AnchorError)) return err;
  // Credit vault errors are 6000-range; reputation errors overlap but
  // come from a different program — the AnchorError's program field
  // tells us which is throwing.
  switch (err.error.errorCode.number) {
    case 6001:
      return new ZeroAmountError();
    case 6002:
      return new VaultFrozenError();
    case 6003:
      return new NotWhitelistedError(ctx.service.toBase58());
    case 6004:
      return new PerTxLimitExceededError(ctx.amountUsdc, ctx.policy.per_tx_limit_usdc);
    case 6005:
      return new HourlyLimitExceededError(
        ctx.amountUsdc,
        ctx.policy.hourly_window_spent_usdc,
        ctx.policy.hourly_limit_usdc,
      );
    case 6006:
      return new LifetimeLimitExceededError(
        ctx.amountUsdc,
        ctx.vault.total_spent,
        ctx.policy.lifetime_limit_usdc,
      );
    // Reputation errors:
    case 6000:
      return new RecordPaymentError("counter overflow on chain");
    case 6011: // ServiceInactive in reputation program
      return new ServiceInactiveError(ctx.service.toBase58());
    default:
      return err;
  }
}

File: clients/sdk-python/src/agent_fuel/pay.py
"""Atomic spend + record_payment + compute_score in a single transaction.

Mirrors `pay()` in clients/sdk/src/pay.ts and `Spender::pay()` in
clients/runtime/src/lib.rs. The agent signs the spend half; the service
keypair co-signs the reputation half; compute_score is bundled so the
on-chain score moves in the same tx (see project_score_pipeline.md)."""

from __future__ import annotations

from dataclasses import dataclass

import httpx
from solders.keypair import Keypair
from solders.pubkey import Pubkey
from solders.signature import Signature

from .anchor_errors import map_pay_error
from .errors import AccountNotFoundError, ZeroAmountError
from .guardrails import guard_spend
from .instructions import (
    compute_score_ix,
    record_payment_ix,
    spend_ix,
)
from .pda import (
    Pubkeyish,
    agent_profile_pda,
    agent_service_link_pda,
    policy_pda,
    receipt_used_pda,
    service_registry_pda,
    to_pubkey,
    vault_pda,
)
from .rpc import _rpc_call, send_and_confirm  # noqa: F401 (re-used by client)
from .spl_token import (
    create_associated_token_account_idempotent_instruction,
    get_associated_token_address,
)
from .types import decode_credit_vault, decode_spend_policy


@dataclass(frozen=True)
class PayResult:
    signature: Signature


async def pay(
    *,
    agent: Keypair,
    service: Keypair,
    owner: Pubkeyish,
    amount_usdc: int,
    receipt_hash: bytes,
    http_client: httpx.AsyncClient,
    rpc_url: str,
) -> PayResult:
    if amount_usdc <= 0:
        raise ZeroAmountError()
    if len(receipt_hash) != 32:
        raise ValueError(f"receipt_hash must be 32 bytes (got {len(receipt_hash)})")

    owner_pk = to_pubkey(owner)
    vault_addr = vault_pda(owner_pk, agent.pubkey())
    policy_addr = policy_pda(vault_addr)

    raw_vault = await _fetch_account_data(http_client, rpc_url, vault_addr)
    raw_policy = await _fetch_account_data(http_client, rpc_url, policy_addr)
    current_slot = await _get_slot(http_client, rpc_url)

    vault = decode_credit_vault(vault_addr, raw_vault)
    policy = decode_spend_policy(policy_addr, raw_policy)
    guard_spend(
        vault=vault,
        policy=policy,
        service=service.pubkey(),
        amount_usdc=amount_usdc,
        current_slot=current_slot,
    )

    service_token_account = get_associated_token_address(vault.usdc_mint, service.pubkey())
    create_ata = create_associated_token_account_idempotent_instruction(
        payer=agent.pubkey(),
        associated_token=service_token_account,
        owner=service.pubkey(),
        mint=vault.usdc_mint,
    )

    profile = agent_profile_pda(agent.pubkey())
    registry = service_registry_pda(service.pubkey())
    link = agent_service_link_pda(profile, registry)
    receipt = receipt_used_pda(receipt_hash)

    ixs = [
        create_ata,
        spend_ix(
            agent=agent.pubkey(),
            vault=vault_addr,
            policy=policy_addr,
            vault_token_account=vault.vault_token_account,
            service_token_account=service_token_account,
            amount_usdc=amount_usdc,
        ),
        record_payment_ix(
            service=service.pubkey(),
            agent_profile=profile,
            service_registry=registry,
            agent_service_link=link,
            receipt_used=receipt,
            amount_usdc=amount_usdc,
            receipt_hash=receipt_hash,
        ),
        compute_score_ix(caller=agent.pubkey(), agent_profile=profile),
    ]
    try:
        sig = await send_and_confirm(http_client, rpc_url, ixs, [agent, service])
    except Exception as e:
        raise map_pay_error(
            e,
            service=service.pubkey(),
            amount_usdc=amount_usdc,
            receipt_hash=receipt_hash,
            vault=vault,
            policy=policy,
        ) from e
    return PayResult(signature=sig)


# --- Internals (also reused by request_spend / register_service) ----------

async def _fetch_account_data(
    client: httpx.AsyncClient, rpc_url: str, address: Pubkey
) -> bytes:
    """Bare `getAccountInfo` returning raw account data. Doesn't enforce
    program ownership — `pay()` already knows which program owns each PDA
    and the on-chain instructions will reject a confused-deputy."""
    import base64

    result = await _rpc_call(
        client,
        rpc_url,
        "getAccountInfo",
        [str(address), {"encoding": "base64", "commitment": "confirmed"}],
    )
    if result is None or result.get("value") is None:
        raise AccountNotFoundError(str(address))
    data_field = result["value"]["data"]
    if isinstance(data_field, list):
        b64 = data_field[0]
    else:
        b64 = data_field
    return base64.b64decode(b64)


async def _get_slot(client: httpx.AsyncClient, rpc_url: str) -> int:
    result = await _rpc_call(client, rpc_url, "getSlot", [{"commitment": "confirmed"}])
    return int(result)
